crates/circuit/src/dag_circuit.rs

```
@@ -0,0 +1,6474 @@
+    ///         ``DAGCircuit``) and will raise a ``DAGCircuitError`` if a cycle
+    ///         would be introduced. This checking comes with a run time
+    ///         penalty. If you can guarantee that your input ``node_block`` is
+    ///         a contiguous block and won't introduce a cycle when it's
+    ///         contracted to a single node, this can be set to ``False`` to
+    ///         improve the runtime performance of this method.
+    ///
+    /// Raises:
+    ///     DAGCircuitError: if ``cycle_check`` is set to ``True`` and replacing
+    ///         the specified block introduces a cycle or if ``node_block`` is
+    ///         empty.
+    ///
+    /// Returns:
+    ///     DAGOpNode: The op node that replaces the block.
+    #[pyo3(signature = (node_block, op, wire_pos_map, cycle_check=true))]
+    fn replace_block_with_op(
+        &mut self,
+        py: Python,
+        node_block: Vec<PyRef<DAGNode>>,
+        op: Bound<PyAny>,
+        wire_pos_map: &Bound<PyDict>,
+        cycle_check: bool,
+    ) -> PyResult<Py<PyAny>> {
+        // If node block is empty return early
+        if node_block.is_empty() {
+            return Err(DAGCircuitError::new_err(
+                "Can't replace an empty 'node_block'",
+            ));
+        }
+
+        let mut qubit_pos_map: HashMap<Qubit, usize> = HashMap::new();
+        let mut clbit_pos_map: HashMap<Clbit, usize> = HashMap::new();
+        for (bit, index) in wire_pos_map.iter() {
+            if bit.is_instance(imports::QUBIT.get_bound(py))? {
+                qubit_pos_map.insert(self.qubits.find(&bit).unwrap(), index.extract()?);
+            } else if bit.is_instance(imports::CLBIT.get_bound(py))? {
+                clbit_pos_map.insert(self.clbits.find(&bit).unwrap(), index.extract()?);
+            } else {
+                return Err(DAGCircuitError::new_err(
+                    "Wire map keys must be Qubit or Clbit instances.",
+                ));
+            }
+        }
```

The Python code does not have this error condition - it silently ignores additional objects (`Var`, for example, which _is_ a valid wire type, but would never have a valid position in the map).  I'm ok including the new failure mode (since the typing of the variables implies that it should only have `Bit` instances), just flagging that this is a change.